### PR TITLE
Feature/175245758 fetch top 5 cell type marker genes

### PR DESCRIPTION
**Pull request by upendrakumbham**
This PR contains code to fetch top5 marker genes.

### CODE REVIEW

**Review by lingyun1010 — COMMENTED** [review]
I tried to make a `GET` request by `http://localhost:8080/gxa/sc/experiments/E-MTAB-6505/marker-genes/cell-types?organismPart=inferred%20cell%20type`, but I got an error in the `MarkerGenesDao` class, given

 ```"PreparedStatementCallback; bad SQL grammar [SELECT g.experiment_accession, m.gene_id, g.variable as cell_group_type, h.value as cell_group_value_where_marker, g.value as cell_group_value, m.marker_probability as marker_p_value, s.mean_expression, s.median_expression FROM scxa_cell_group_marker_gene_stats s, scxa_cell_group_marker_genes m, scxa_cell_group g, scxa_cell_group h WHERE s.cell_group_id=g.id and s.marker_id=m.id and m.cell_group_id = h.id and g.experiment_accession= ? and m.marker_probability < 0.05 and g.variable = ? and g.value IN (?, ?) and expression_type=0 order by m.marker_probability ]; nested exception is org.postgresql.util.PSQLException: ERROR: relation \"scxa_cell_group_marker_gene_stats\" does not exist\n  Position: 229"```,

while I have a good connection with the PostgreSQL database. Would you please give a go as well, if you could get the JSON response?

File: src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGeneServiceImpl.java [hunk 1]
```diff
@@ -20,8 +22,14 @@ public MarkerGeneServiceImpl(MarkerGenesDao markerGenesDao) {
...
      */
     @Override
-    public ImmutableList<CellTypeMarkerGene> getCellTypeMarkerGeneProfile(String experimentAccession, String organismPart) {
-        List<CellTypeMarkerGene> cellTypeMarkerGenes = markerGenesDao.getCellTypeMarkerGenes(experimentAccession, organismPart);
-        return ImmutableList.copyOf(cellTypeMarkerGenes);
+    public Map<String, ImmutableSet<CellTypeMarkerGene>> getCellTypeMarkerGeneProfile(String experimentAccession, String organismPart) {
+        return markerGenesDao.getCellTypeMarkerGenes(experimentAccession, organismPart)
+                .stream()
+                .collect(groupingBy(CellTypeMarkerGene::cellGroupValue, toImmutableSet())).entrySet().stream()
+                .collect(toMap(cellTypeMarkerGenes -> cellTypeMarkerGenes.getKey(),
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGeneServiceImpl.java:29]: Use `toImmutableMap` and return type accordingly.

File: src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/HighchartsHeatmapAdapter.java [hunk 2]
```diff
@@ -77,22 +79,28 @@ public HighchartsHeatmapAdapter(BioEntityPropertyDao bioEntityPropertyDao) {
...
+        List<CellTypeMarkerGene> allMarkerGenes = markerGenesMap.values().stream()
+                                                    .flatMap(Collection::stream)
+                                                    .collect(Collectors.toList());
+
+        var sortedMarkerGenes = allMarkerGenes.stream()
+                                .parallel()
+                                .sorted(comparing(CellTypeMarkerGene::cellGroupValueWhereMarker).thenComparing(CellTypeMarkerGene::pValue))
+                                .collect(toImmutableList());
+
+        var rows = sortedMarkerGenes.stream()
```
Comment (alfonsomunozpomer) [inline src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/HighchartsHeatmapAdapter.java:98]: You can chain these two blocks, right?

**Review by alfonsomunozpomer — CHANGES_REQUESTED** [review]
All good and simple. Just a test that checks that the number of retrieved marker genes is at most 5 is missing.

File: src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/HighchartsHeatmapAdapter.java (hunk 2 above)
Comment (upendrakumbham) [reply src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/HighchartsHeatmapAdapter.java:98]: Yes, we can. I want to get rid of this Branch, as I duplicated the same code in the [Update cluster PR](https://github.com/ebi-gene-expression-group/atlas-web-single-cell/pull/113). I will fix these comments in the Update cluster PR.

File: src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGeneServiceImpl.java (hunk 1 above)
Comment (upendrakumbham) [reply src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGeneServiceImpl.java:29]: Fixed.

Comment (upendrakumbham) [reply src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/MarkerGeneServiceImpl.java:29]: I will add this missing test case for the top5 genes in the [Update cluster PR](https://github.com/ebi-gene-expression-group/atlas-web-single-cell/pull/113).

**Pull request closed without merging**